File: mclaude-web/src/lib/slug.test.ts
import { describe, it, expect } from 'vitest'
import { slugify, validate, deriveUserSlug, RESERVED_WORDS } from './slug'

describe('slugify', () => {
  it('lowercases ASCII', () => {
    expect(slugify('Hello World')).toBe('hello-world')
  })

  it('strips accents via NFD decomposition', () => {
    expect(slugify('Caf\u00e9')).toBe('cafe')
    expect(slugify('Na\u00efve')).toBe('naive')
    expect(slugify('r\u00e9sum\u00e9')).toBe('resume')
  })

  it('handles emoji-only input (returns empty string)', () => {
    // Emoji get stripped after NFD + combining mark removal, leaving only hyphens
    // which are then trimmed
    const result = slugify('\u{1F600}')
    expect(result).toBe('')
  })

  it('replaces runs of non-alnum with a single hyphen', () => {
    expect(slugify('foo  bar---baz')).toBe('foo-bar-baz')
    expect(slugify('hello!@#world')).toBe('hello-world')
  })

  it('trims leading and trailing hyphens', () => {
    expect(slugify('---hello---')).toBe('hello')
    expect(slugify('   spaces   ')).toBe('spaces')
  })

  it('truncates at 63 characters', () => {
    const long = 'a'.repeat(100)
    expect(slugify(long)).toHaveLength(63)
  })

  it('does not end with a hyphen after truncation', () => {
    // 62 'a' chars + non-alnum run — truncation should not leave trailing hyphen
    const s = 'a'.repeat(62) + '---more'
    const result = slugify(s)
    expect(result).not.toMatch(/-$/)
  })

  it('handles empty string', () => {
    expect(slugify('')).toBe('')
  })

  it('handles numeric-only names', () => {
    expect(slugify('42')).toBe('42')
  })
})

describe('validate', () => {
  it('accepts a valid slug', () => {
    expect(validate('alice-gmail')).toBeNull()
    expect(validate('my-project-123')).toBeNull()
    expect(validate('a')).toBeNull()
  })

  it('rejects empty string', () => {
    expect(validate('')?.message).toContain('empty')
  })

  it('rejects leading underscore', () => {
    expect(validate('_internal')).not.toBeNull()
  })

  it('rejects uppercase', () => {
    expect(validate('Alice')).not.toBeNull()
  })

  it('rejects spaces', () => {
    expect(validate('hello world')).not.toBeNull()
  })

  it('rejects slug longer than 63 chars', () => {
    expect(validate('a'.repeat(64))).not.toBeNull()
  })

  it('accepts slug of exactly 63 chars', () => {
    expect(validate('a'.repeat(63))).toBeNull()
  })

  it('rejects all 10 reserved words', () => {
    for (const word of RESERVED_WORDS) {
      const err = validate(word)
      expect(err, `reserved word "${word}" should be rejected`).not.toBeNull()
    }
  })

  it('rejects slug starting with hyphen', () => {
    expect(validate('-hello')).not.toBeNull()
  })

  it('accepts slug starting with digit', () => {
    expect(validate('1project')).toBeNull()
  })

  it('rejects slug with dot', () => {
    expect(validate('hello.world')).not.toBeNull()
  })

  it('rejects slug with slash', () => {
    expect(validate('hello/world')).not.toBeNull()
  })
})

describe('deriveUserSlug', () => {
  it('combines slugified name with domain first segment', () => {
    expect(deriveUserSlug('Richard', 'richard@rbc.com')).toBe('richard-rbc')
    expect(deriveUserSlug('Alice', 'alice@gmail.com')).toBe('alice-gmail')
  })

  it('ADR-0024 example: richard@rbc.com', () => {
    expect(deriveUserSlug('Richard Song', 'richard@rbc.com')).toBe('richard-song-rbc')
  })

  it('ADR-0024 example: alice@gmail.com', () => {
    expect(deriveUserSlug('Alice', 'alice@gmail.com')).toBe('alice-gmail')
  })

  it('uses only domain first segment for multi-segment domains', () => {
    // bob@company.co.uk → bob-company (not bob-company-co-uk)
    expect(deriveUserSlug('Bob', 'bob@company.co.uk')).toBe('bob-company')
  })

  it('falls back to email local-part when display name is empty', () => {
    expect(deriveUserSlug('', 'alice@gmail.com')).toBe('alice-gmail')
  })

  it('handles email without @ domain part', () => {
    // No domain segment — just use the name/local part
    const result = deriveUserSlug('Alice', 'alice')
    expect(result).toBe('alice')
  })

  it('truncates combined slug to 63 chars', () => {
    const longName = 'a'.repeat(60)
    const result = deriveUserSlug(longName, 'user@example.com')
    expect(result.length).toBeLessThanOrEqual(63)
  })

  it('does not end with hyphen', () => {
    const result = deriveUserSlug('test', 'test@x.com')
    expect(result).not.toMatch(/-$/)
  })
})

File: mclaude-common/pkg/slug/slug_test.go
package slug_test

import (
	"strings"
	"testing"

	"mclaude-common/pkg/slug"
)

// --------------------------------------------------------------------------
// Slugify tests
// --------------------------------------------------------------------------

func TestSlugify(t *testing.T) {
	cases := []struct {
		name  string
		input string
		want  string
	}{
		{
			name:  "pure ASCII lowercase",
			input: "hello-world",
			want:  "hello-world",
		},
		{
			name:  "pure ASCII with spaces",
			input: "my project name",
			want:  "my-project-name",
		},
		{
			name:  "accented characters é",
			input: "café",
			want:  "cafe",
		},
		{
			name:  "accented character ñ",
			input: "niño",
			want:  "nino",
		},
		{
			name:  "mixed accents",
			input: "résumé",
			want:  "resume",
		},
		{
			name:  "emoji-only input",
			input: "🎉",
			want:  "", // all non-[a-z0-9] stripped, result is empty after trim
		},
		{
			name:  "leading punctuation",
			input: "---hello",
			want:  "hello",
		},
		{
			name:  "trailing punctuation",
			input: "hello---",
			want:  "hello",
		},
		{
			name:  "too-long string",
			input: strings.Repeat("a", 100),
			want:  strings.Repeat("a", 63),
		},
		{
			name:  "reserved word 'users'",
			input: "users",
			want:  "users", // Slugify itself does not reject reserved words; caller uses ValidateOrFallback
		},
		{
			name:  "empty string",
			input: "",
			want:  "",
		},
		{
			name:  "uppercase",
			input: "Richard Song",
			want:  "richard-song",
		},
		{
			name:  "mixed punctuation",
			input: "hello.world/foo bar",
			want:  "hello-world-foo-bar",
		},
		{
			name:  "leading underscore stripped to dash then trimmed",
			input: "_internal",
			want:  "internal",
		},
		{
			name:  "digits preserved",
			input: "project-42",
			want:  "project-42",
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := slug.Slugify(tc.input)
			if got != tc.want {
				t.Errorf("Slugify(%q) = %q, want %q", tc.input, got, tc.want)
			}
		})
	}
}

// --------------------------------------------------------------------------
// Validate tests
// --------------------------------------------------------------------------

func TestValidate(t *testing.T) {
	validCases := []struct {
		name  string
		input string
	}{
		{"simple lowercase", "hello"},
		{"with hyphen", "hello-world"},
		{"starts with digit", "3d-printer"},
		{"max length exactly 63", strings.Repeat("a", 63)},
		{"single char", "a"},
	}

	for _, tc := range validCases {
		t.Run("valid/"+tc.name, func(t *testing.T) {
			if err := slug.Validate(tc.input); err != nil {
				t.Errorf("Validate(%q) returned unexpected error: %v", tc.input, err)
			}
		})
	}

	invalidCases := []struct {
		name      string
		input     string
		wantType  string // error type name fragment
	}{
		{"empty string", "", "ErrEmpty"},
		{"leading underscore", "_internal", "ErrLeadingUnderscore"},
		{"contains uppercase", "Hello", "ErrCharset"},
		{"contains dot", "hello.world", "ErrCharset"},
		{"contains space", "hello world", "ErrCharset"},
		{"starts with hyphen", "-hello", "ErrCharset"},
		{"too long (64 chars)", strings.Repeat("a", 64), "ErrTooLong"},
		{"reserved: users", "users", "ErrReserved"},
		{"reserved: hosts", "hosts", "ErrReserved"},
		{"reserved: projects", "projects", "ErrReserved"},
		{"reserved: sessions", "sessions", "ErrReserved"},
		{"reserved: clusters", "clusters", "ErrReserved"},
		{"reserved: api", "api", "ErrReserved"},
		{"reserved: events", "events", "ErrReserved"},
		{"reserved: lifecycle", "lifecycle", "ErrReserved"},
		{"reserved: quota", "quota", "ErrReserved"},
		{"reserved: terminal", "terminal", "ErrReserved"},
		{"contains slash", "hello/world", "ErrCharset"},
		{"contains wildcard *", "hello*", "ErrCharset"},
		{"contains >", "hello>", "ErrCharset"},
	}

	for _, tc := range invalidCases {
		t.Run("invalid/"+tc.name, func(t *testing.T) {
			err := slug.Validate(tc.input)
			if err == nil {
				t.Errorf("Validate(%q) expected error, got nil", tc.input)
			}
		})
	}
}

// --------------------------------------------------------------------------
// ValidateOrFallback tests
// --------------------------------------------------------------------------

func TestValidateOrFallback(t *testing.T) {
	seed := [16]byte{0xde, 0xad, 0xbe, 0xef, 0x01, 0x02, 0x03, 0x04}

	t.Run("valid candidate is returned unchanged", func(t *testing.T) {
		got := slug.ValidateOrFallback("my-project", slug.KindProject, seed)
		if got != "my-project" {
			t.Errorf("want %q, got %q", "my-project", got)
		}
	})

	t.Run("empty candidate triggers fallback", func(t *testing.T) {
		got := slug.ValidateOrFallback("", slug.KindUser, seed)
		if !strings.HasPrefix(got, "u-") {
			t.Errorf("expected fallback to start with 'u-', got %q", got)
		}
		if err := slug.Validate(got); err != nil {
			t.Errorf("fallback slug %q is not valid: %v", got, err)
		}
	})

	t.Run("reserved word triggers fallback for project", func(t *testing.T) {
		got := slug.ValidateOrFallback("projects", slug.KindProject, seed)
		if !strings.HasPrefix(got, "p-") {
			t.Errorf("expected fallback to start with 'p-', got %q", got)
		}
		if err := slug.Validate(got); err != nil {
			t.Errorf("fallback slug %q is not valid: %v", got, err)
		}
	})

	t.Run("emoji-only triggers fallback for session", func(t *testing.T) {
		// Slugify("🎉") == "" — ValidateOrFallback receives that empty string
		candidate := slug.Slugify("🎉")
		got := slug.ValidateOrFallback(candidate, slug.KindSession, seed)
		if !strings.HasPrefix(got, "s-") {
			t.Errorf("expected fallback to start with 's-', got %q", got)
		}
		if err := slug.Validate(got); err != nil {
			t.Errorf("fallback slug %q is not valid: %v", got, err)
		}
	})

	t.Run("fallback is deterministic", func(t *testing.T) {
		a := slug.ValidateOrFallback("", slug.KindCluster, seed)
		b := slug.ValidateOrFallback("", slug.KindCluster, seed)
		if a != b {
			t.Errorf("expected identical fallbacks, got %q and %q", a, b)
		}
	})

	t.Run("different seeds produce different fallbacks", func(t *testing.T) {
		seed2 := [16]byte{0x00, 0x00, 0x00, 0x01}
		a := slug.ValidateOrFallback("", slug.KindUser, seed)
		b := slug.ValidateOrFallback("", slug.KindUser, seed2)
		if a == b {
			t.Errorf("expected different fallbacks for different seeds, both got %q", a)
		}
	})

	t.Run("host kind produces h- prefix", func(t *testing.T) {
		got := slug.ValidateOrFallback("", slug.KindHost, seed)
		if !strings.HasPrefix(got, "h-") {
			t.Errorf("expected fallback to start with 'h-', got %q", got)
		}
	})
}

// --------------------------------------------------------------------------
// Round-trip: Slugify → Validate
// --------------------------------------------------------------------------

func TestSlugifyRoundTrip(t *testing.T) {
	// Any non-empty result from Slugify must be valid (except reserved words,
	// which Slugify passes through — those need ValidateOrFallback).
	inputs := []string{
		"hello world",
		"Richard Song",
		"café au lait",
		"123-test",
		"My Project Name",
		"a",
		strings.Repeat("b", 200),
	}
	for _, input := range inputs {
		got := slug.Slugify(input)
		if got == "" {
			continue // empty results need ValidateOrFallback, not Validate
		}
		if err := slug.Validate(got); err != nil {
			// Reserved-word results are expected to fail Validate — that is the
			// intended contract: callers use ValidateOrFallback.
			if _, ok := err.(slug.ErrReserved); ok {
				continue
			}
			t.Errorf("Validate(Slugify(%q)) = %q, got error: %v", input, got, err)
		}
	}
}

// --------------------------------------------------------------------------
// DeriveUserSlug tests
// --------------------------------------------------------------------------

func TestDeriveUserSlug(t *testing.T) {
	cases := []struct {
		name        string
		displayName string
		email       string
		want        string
	}{
		{
			name:        "standard user",
			displayName: "Richard",
			email:       "richard@rbc.com",
			want:        "richard-rbc",
		},
		{
			name:        "multi-part name",
			displayName: "Richard Song",
			email:       "richard@rbc.com",
			want:        "richard-song-rbc",
		},
		{
			name:        "alice gmail",
			displayName: "Alice",
			email:       "alice@gmail.com",
			want:        "alice-gmail",
		},
		{
			name:        "domain with multiple segments — first segment only",
			displayName: "user",
			email:       "user@rbc.co.uk",
			want:        "user-rbc",
		},
		{
			name:        "no display name — use email local-part",
			displayName: "",
			email:       "alice@gmail.com",
			want:        "alice-gmail",
		},
		{
			name:        "display name overrides local-part",
			displayName: "Bob",
			email:       "robert@company.org",
			want:        "bob-company",
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := slug.DeriveUserSlug(tc.displayName, tc.email)
			if got != tc.want {
				t.Errorf("DeriveUserSlug(%q, %q) = %q, want %q",
					tc.displayName, tc.email, got, tc.want)
			}
		})
	}
}

// --------------------------------------------------------------------------
// MustParse panic tests
// --------------------------------------------------------------------------

func TestMustParseUserSlugPanics(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Error("expected panic for invalid slug, got none")
		}
	}()
	slug.MustParseUserSlug("invalid slug!")
}

func TestMustParseUserSlugValid(t *testing.T) {
	s := slug.MustParseUserSlug("alice-gmail")
	if s != "alice-gmail" {
		t.Errorf("expected %q, got %q", "alice-gmail", s)
	}
}
